Fix object layout races

daumayr opened the pull request:
This PR fixes issue #85.

The Phaser used in ObjectTransitionSafepoint used a default Phaser, which Terminates when onAdvance(...) returns true. This caused the synchronisation methods to have no effect.
The default onAdvance implementation causes the Phaser to terminate when the number of registered Threads is 0.

By overwriting the onAdvance method and returning false we can prevent Phaser from terminating and preserve the synchronisation.

daumayr committed:
Fix ObjectTransitionSafepoint

Phaser was able to terminate, an thread registration didn't prevent threads from making progress.

daumayr committed:
Fix ClassSlot initialisation Deadlock

daumayr committed:
Replace safepoint assumption at end

daumayr committed:
Fix registration loophole

daumayr committed:
Add Missing Unsafe Method

smarr commented on src/som/interpreter/objectstorage/ObjectTransitionSafepoint.java:
```diff
@@ -10,6 +10,7 @@
 import com.oracle.truffle.api.nodes.InvalidAssumptionException;
 
 import som.compiler.MixinDefinition.SlotDefinition;
+import som.vm.CustomPhaser;
```
CustomPhaser, not a great name.
How about: `SafepointPhaser`?

daumayr replied:
good idea

smarr commented on src/som/vmobjects/SObject.java:
```diff
@@ -517,4 +517,22 @@ private void setFieldAfterLayoutChange(final SlotDefinition slot,
+    return loc instanceof ObjectStorageLocation;
+  }
+
+  public final synchronized void ensureSlotAllocatedToAvoidDeadlock(
```
Hmmmmmm. So, we still hold a lock when we may trigger a safepoint? Any way around that?

Seems like this is only moving the deadlock?

daumayr replied:
this method is called after the safepoint starts, and nowhere else. the synchronized is for the case that two threads enter the safepoint to update the same object.

smarr replied:
Aaaaah, right, now I see it.

Should have looked at it in a proper IDE to start with :)
Thanks!

daumayr commented:
While tests are running for this branch, we still have an issue with some benchmarks. The Vacation Benchmark still deadlocks, so ideally we don't merge this PR until we fix the Join Prim and similar things.

smarr commented:
Ok, will see what I can do about those. Will probably try to do them in a separate PR to see performance issues independently.

daumayr commented:
I think we need to exclude those benchmarks (LeeTm and Vacation) until then.

daumayr committed:
Add fix to changelog

daumayr committed:
Remove unnecessary whitespace

smarr commented on src/som/interpreter/nodes/dispatch/ClassSlotAccessNode.java:
```diff
@@ -60,6 +61,11 @@ public SClass read(final SObject rcvr) {
       return (SClass) cachedValue;
     }
 
+    if (!rcvr.isObjectSlotAllocated(slotDef)) {
```
Does this test compile out? We are here at a point where we are guaranteed to have one specific layout (which our guard tested before).
We may want to test against that instead of the `rcvr` object, which may have changed in the mean time.

daumayr replied:
not sure if this compiles away, after all the rcvr isn't final. Probably would be better to use the guards layout, and maybe an assumption that get's invalidated when the slot is allocated.

smarr replied:
This issue is now resolved with: https://github.com/smarr/SOMns/pull/285/commits/88b3b2b9a4c918c731b0a4fe6e6c951f41a371a9

Replacing a hash lookup with a constant :)

smarr commented on src/som/interpreter/objectstorage/ObjectTransitionSafepoint.java:
```diff
@@ -140,12 +155,27 @@ public void writeAndGeneralizeSlot(final SObject obj, final SlotDefinition slot,
+    int active = transitionsInProgress.decrementAndGet();
+
+    if (active == 0) {
+      synchronized (this) {
```
Is it on purpose that you got two separate synchronization mechanisms here?

This might be an optimization I guess, but I may be simpler to stick to exactly one (`synchronized` only?), if possible?

daumayr replied:
are you referring to the atomic int? I suppose we can change it to a normal int, and move all operations on it inside the synchronized. A synchronized then also needs to be introduced into the waitForSafepointStart.

smarr replied:
Addressed with: https://github.com/smarr/SOMns/pull/285/commits/33f924f00754b362bd1ae423d52cf3f15b8df155

This way, there is one uniform mechanism, and performance is hopefully not a major issue. Though, not really measured yet.

smarr reviewed (commented):
A quick review to give at least some comments.
Sorry, didn't get to more.

smarr committed:
Added SafepointPhaser based on JRE-166 Java 8.  The source of the CustomPhaser is   http://gee.cs.oswego.edu/cgi-bin/viewcvs.cgi/jsr166/src/jdk8/java/util/concurrent/Phaser.java?revision=1.3

This is a simple copy, no modifications yet.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adapt to be SafepointPhaser

This change introduces our formatting and names the class SafepointPhaser and sets the package name correctly.

No changes of functionality.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix checkstyle issues in SafepointPhaser

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use SafepointPhaser

Co-authored-by: Dominik Aumayr <daumayr162@gmail.com>
Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove termination by onAdvance from SafepointPhaser

The SafepointPhaser should not terminate, even if no parties are registered on it anymore.

This can happen with our safepoint design, but it is not safe to terminate the phaser. It needs to continue working and accept threads to join again.

Co-authored-by: Dominik Aumayr <daumayr162@gmail.com>
Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Do not use managed blocking for SafepointPhaser

The phaser is intended for use in safepoints, where managed blocking is counter productive.

- remove ManagedBlocker implementation from SafepointPhaser

Co-authored-by: Dominik Aumayr <daumayr162@gmail.com>
Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Simplify SafepointPhaser

- we don’t need support for a phaser tree
- we don’t need different constructors

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove further unused elements from SafepointPhaser

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use available Unsafe object from TraceBuffer

Co-authored-by: Dominik Aumayr <daumayr162@gmail.com>
Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Simplify SafepointPhaser QNode to needed features

Remove interruptible, timed, and deadline notions.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make SafepointPhaser final and restrict visibility

SafepointPhaser isn’t a general purpose class, so, let’s provide features only when needed.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Determine on ClassSlotAccessNode instantiation whether slot is allocated

This makes the run-time check compile away.
Slot allocation is still lazy, i.e., on use.

This relies on the general mechanism in the dispatch chain to remove entries with invalid guards.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Integrate SafepointPhaser closely with Safepoint to avoid races

This change uses the knowledge of the safepoint consisting of two phases. This allows us to manage registrations without extra race, and it allows us to renew the assumption as part of the arrival routine in the phaser.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented on src/som/interpreter/objectstorage/ObjectTransitionSafepoint.java:
```diff
@@ -59,6 +60,19 @@ private static Assumption create() {
   @TruffleBoundary
   public void register() {
     phaser.register();
+    try {
```
@daumayr I am not sure I am getting the full logic here.

We can have the following situations:
- register while no safepoint is done
- register while safepoint is starting
- register while safepoint is in progress

Can we distinguish the last two? only by looking at `transitionsInProgress`, no?

smarr committed:
Reduce SafepointPhaser documentation to the accurate bits

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@daumayr I come up with what I think is a working solution. Could you please review and possible test this?
I haven't yet thought of a good test on how to stress the safepoints.
Not sure I can induce repeated safepoints easily.
Would be good to discuss this in detail, to see whether there are any know races left, or any other issue that may make this approach problematic.

The most important bit to discuss is likely https://github.com/smarr/SOMns/pull/285/commits/0274546403810d7f522e6f1ae8a8b67b8d9f2b6e

daumayr commented:
`(phase & 1) == 0`
I was thinking of using something similar when i did some work on the safepoints. When no safepoint is going on and all threads deregister the phaser advances to the next phase. After that the even/odd check would no longer work. You might want to change the doArrive method to not advance phase when there are no threads left.

smarr commented:
@daumayr hmmmm, ok. Didn't think of that. Looking at the code: https://github.com/smarr/SOMns/pull/285/files#diff-2b3be932b680eb719ab91a02f3b6f064R137

I suppose, I could simply increment the phase in that case by 2, no?

smarr commented:
hmmm, well, possibly depending on in which concrete state we are. Though, we should never unregister during a safepoint. So, probably increment by 2 plus an assertion that we are never in a phase that indicates a safepoint is going on

daumayr commented:
Increment by two sounds good, that way we still have a phase change and keep the odd/even.

smarr commented:
ok, cool.

and one more todo:

- [x] add unit tests that test the basic logic for correctness, and have asserts for all the parts not covered

smarr committed:
SafepointPhase: ensure odd/even for phases also after all threads are reregistered

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added basic SafepointPhaser unit test

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@daumayr could you give this a spin with your snapshot tests?
Would also be great if you could review it and look at the added test. Any obvious test we should add?

daumayr commented:
Snapshot Tests running, didn't see any errors.

daumayr commented on tests/java/som/interpreter/objectstorage/SafepointPhaserTest.java:
```diff
@@ -0,0 +1,96 @@
+    instanceSlots.add(slotDef);
+
+    instanceClass = new SClass(null);
+    factory = new ClassFactory(null, null, instanceSlots, null, false, false, false, null,
```
This seems to have a wrong number of arguments on my machine.

smarr replied:
Hm, what exactly do you mean with this?
Also, is this without any other changes? Seems to compile fine on Travis and Brussels.
